ext/storage_sources/s3_store/s3_connection.cpp

```diff
@@ -13,51 +13,74 @@
 
 /*
  * ListBuckets --
- *     Builds a list of buckets from AWS account into a vector. Returns true if success, otherwise
- *     false.
+ *     Builds a list of buckets from AWS account into a vector. Returns 0 if success, otherwise
+ *     1.
  */
-bool
+int
 S3Connection::ListBuckets(std::vector<std::string> &buckets) const
 {
-    auto outcome = m_S3CrtClient.ListBuckets();
+    auto outcome = s3CrtClient.ListBuckets();
     if (outcome.IsSuccess()) {
         for (const auto &bucket : outcome.GetResult().GetBuckets())
             buckets.push_back(bucket.GetName());
-        return true;
+        return (0);
     } else {
         std::cerr << "Error in ListBuckets: " << outcome.GetError().GetMessage() << std::endl
                   << std::endl;
-        return false;
+        return (1);
     }
 }
 
 /*
  * ListObjects --
- *     Builds a list of object names from a S3 bucket into a vector. Returns true if success,
- *     otherwise false.
+ *     Builds a list of object names, with prefix matching, from an S3 bucket into a vector. The
+ *     batchSize parameter specifies the maximum number of objects returned in each AWS response, up
+ *     to 1000. Returns 0 if success, otherwise 1.
  */
-bool
-S3Connection::ListObjects(const std::string &bucketName, std::vector<std::string> &objects) const
+int
+S3Connection::ListObjects(const std::string &bucketName, const std::string &prefix,
+  std::vector<std::string> &objects, uint32_t batchSize, bool listSingle) const
 {
-    Aws::S3Crt::Model::ListObjectsRequest request;
-    request.WithBucket(bucketName);
-    Aws::S3Crt::Model::ListObjectsOutcome outcomes = m_S3CrtClient.ListObjects(request);
+    Aws::S3Crt::Model::ListObjectsV2Request request;
+    request.SetBucket(bucketName);
+    request.SetPrefix(prefix);
+
+    if (listSingle)
+        request.SetMaxKeys(1);
+    else
+        request.SetMaxKeys(batchSize);
```

Comment:
What happens if the provided batch size is more than 1000, I think we should fail the function with EINVAL or similar failure in such a case.

Reply:
AWS sets the value to 1000. I don't think we need to fail the function.